### PR TITLE
PR_CRP_CheckForDP

### PR DESCRIPTION
**Pull request by echardon**
The DP team added a sanity check verifying that only 2D clusters that have hits, at least partially, in a common daughter volume should be matched.

We added some Geometry and Cluster helper functions to perform the check.

It is done in three algorithms : 1) TwoViewTransverseTrack 2)CosmicRayTrackRecovery**(this one actually touches Single-Phase)** 3)CosmicRayBaseMatching **(this one also touches Single-Phase)**

A small addition was also made in the MasterAlgorithm.cc in order to initialize properly the new LArCaloHit object that now have a new member variable called m_daughterVolumeId.

A parallel pull request https://github.com/PandoraPFA/larpandora/pull/9 has been issued for larpandora repository regarding the implementation of daughter volumes for DP and the addition of the new member variable in LArCaloHit.

### CODE REVIEW

**Comment by echardon** [comment]
This PR is a follow up to https://github.com/PandoraPFA/LArContent/pull/132 . The branch used in the previous PR had an problematic commit history, we figured it was easier to go with a new clean branch.

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc [hunk 1]
```diff
@@ -670,6 +670,7 @@ StatusCode MasterAlgorithm::Copy(const Pandora *const pPandora, const CaloHit *c
     // ATTN Parent of calo hit in worker is corresponding calo hit in master
     parameters.m_pParentAddress = static_cast<const void*>(pCaloHit);
     parameters.m_larTPCVolumeId = pLArCaloHit->GetLArTPCVolumeId();
+    parameters.m_daughterVolumeId = pLArCaloHit->GetDaughterVolumeId();
```
Comment (AndyChappell) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.cc:673]: Looks like this change is a subset of Maria's changes, so can pick Maria's version in any merge.

File: larpandoracontent/LArHelpers/LArClusterHelper.cc [hunk 2]
```diff
@@ -587,6 +588,24 @@ void LArClusterHelper::GetCaloHitListInBoundingBox(const pandora::Cluster *const
...
 
+void LArClusterHelper::GetDaughterVolumeIDs(const Cluster *const pCluster, UIntSet &daughterVolumeIds)
+{
+    const OrderedCaloHitList &orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+
+    for (OrderedCaloHitList::const_iterator ochIter = orderedCaloHitList.begin(), ochIterEnd = orderedCaloHitList.end(); ochIter != ochIterEnd; ++ochIter)
+    {
+        for (CaloHitList::const_iterator hIter = ochIter->second->begin(), hIterEnd = ochIter->second->end(); hIter != hIterEnd; ++hIter)
+        {
+            const CaloHit *const pCaloHit = *hIter;
```
Comment (AndyChappell) [inline larpandoracontent/LArHelpers/LArClusterHelper.cc:599]: Prefer `pCaloHit(*hIter)` over `pCaloHit = *hIter`

File: larpandoracontent/LArHelpers/LArClusterHelper.h [hunk 3]
```diff
@@ -254,6 +257,16 @@ class LArClusterHelper
     static void GetCaloHitListInBoundingBox(const pandora::Cluster *const pCluster, const pandora::CartesianVector &lowerBound, 
         const pandora::CartesianVector &upperBound, pandora::CaloHitList &caloHitList);
 
+    /**
+     *  @brief  Get the set of the daughter volumes that contains the cluster
+     *
+     *  @param  daughterVolumeIds set of daughter volume Ids pCluster lives in
+     *  @param  pCluster1 address of the cluster
+     *
+     *  @return the set of daughter volumes
```
Comment (AndyChappell) [inline larpandoracontent/LArHelpers/LArClusterHelper.h:266]: Swap the order of the params here and remove '1' from pCluster. Would just note that daughterVolumeIds is the output variable here and remove the return description

File: larpandoracontent/LArHelpers/LArGeometryHelper.h [hunk 4]
```diff
@@ -197,6 +200,15 @@ class LArGeometryHelper
      */
     static pandora::CartesianVector GetWireAxis(const pandora::Pandora &pandora, const pandora::HitType view);
 
+    /**
+     *  @brief  Return the set of common daughter volumes between two 2D clusters
+     *
+     *  @param  intersect the set of shared daughter volumes
```
Comment (AndyChappell) [inline larpandoracontent/LArHelpers/LArGeometryHelper.h:206]: Move to last position

File: larpandoracontent/LArObjects/LArCaloHit.h [hunk 5]
```diff
@@ -28,6 +28,7 @@ class LArCaloHitParameters : public object_creation::CaloHit::Parameters
 {
 public:
     pandora::InputUInt      m_larTPCVolumeId;       ///< The lar tpc volume id
+    pandora::InputUInt      m_daughterVolumeId;     ///< The daughter volume id
```
Comment (AndyChappell) [inline larpandoracontent/LArObjects/LArCaloHit.h:31]: These changes appear to be a subset of the changes in Maria's branch. As with the master algorithm, it looks like the simple option here is to pick Maria's changes for this file.

File: larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayTrackRecoveryAlgorithm.cc [hunk 6]
```diff
@@ -196,6 +196,13 @@ void CosmicRayTrackRecoveryAlgorithm::MatchClusters(const Cluster* const pSeedCl
     {
         const Cluster *const pTargetCluster = *tIter;
 
+        UIntSet daughterVolumeIntersection;
+        LArGeometryHelper::GetCommonDaughterVolumes(pSeedCluster, pTargetCluster, daughterVolumeIntersection);
+
+        if (daughterVolumeIntersection.empty())
+            continue;
+
+
```
Comment (AndyChappell) [inline larpandoracontent/LArThreeDReco/LArCosmicRay/CosmicRayTrackRecoveryAlgorithm.cc:205]: Excess whitespace

**Review by AndyChappell — COMMENTED** [review]
Subject to the comments relating to merging with Maria's changes I think this PR looks in good shape. It should be possible to just eliminate the changes in MasterAlgorithm.cc and LArCaloHit.h and rebase against a branch with Maria's changes once those are merged in.

**Pull request closed without merging**